cli_tools/common/utils/daisyutils/daisy_worker.go

```diff
@@ -74,35 +78,93 @@ func createResourceLabelerIfMissing(env EnvironmentSettings, hooks []interface{}
-	logger logging.Logger
-	env    EnvironmentSettings
-	hooks  []interface{}
+	workflowProvider WorkflowProvider
+	finished         *daisy.Workflow
+	logger           logging.Logger
+	env              EnvironmentSettings
+	hooks            []interface{}
+
+	cancel      chan string
+	cancelGuard sync.Once
 }
 
 // Run runs the daisy workflow with the supplied vars.
-func (w *defaultDaisyWorker) Run(vars map[string]string) error {
-	if err := (&ApplyAndValidateVars{w.env, vars}).PreRunHook(w.wf); err != nil {
-		return err
+func (w *defaultDaisyWorker) Run(vars map[string]string) (err error) {
+	var wf *daisy.Workflow
+	for attempts := 0; attempts <= 1; attempts++ {
+		if wf, err = w.workflowProvider(); err != nil {
+			break
+		}
+		if err = w.checkIfCancelled(wf); err != nil {
+			break
+		}
+		var retryRequested bool
+		retryRequested, err = w.runOnce(wf, vars)
+		if err == nil || !retryRequested {
+			break
+		}
+		w.logger.Debug("Retry requested. err=" + err.Error())
+	}
+	w.finished = wf
+	return err
+}
+
+// checkIfCancelled determines whether the workflow has been cancelled internally,
+// or whether a client of DaisyWorker has called cancel. If so, then a non-nil
+// error is returned describing the cancellation.
+func (w *defaultDaisyWorker) checkIfCancelled(wf *daisy.Workflow) (err error) {
+	canceled := false
+	reason := ""
+	select {
+	case reason = <-w.cancel:
+		canceled = true
+	case <-wf.Cancel:
+		canceled = true
+	default:
+		break
+	}
+	if canceled {
+		msg := "workflow canceled"
+		if reason != "" {
+			msg = fmt.Sprintf("%s: %s", msg, reason)
+		}
+		err = errors.New(msg)
+	}
+	return err
+}
+
+// runOnce applies vars to the workflow, runs hooks, and runs the workflow. The retry
+// return value indicates whether a hook has requested a retry.
+func (w *defaultDaisyWorker) runOnce(wf *daisy.Workflow, vars map[string]string) (retry bool, err error) {
+	if err := (&ApplyAndValidateVars{w.env, vars}).PreRunHook(wf); err != nil {
+		return false, err
 	}
 	for _, hook := range w.hooks {
 		preHook, isPreHook := hook.(WorkflowPreHook)
 		if isPreHook {
-			if err := preHook.PreRunHook(w.wf); err != nil {
-				return err
+			if err := preHook.PreRunHook(wf); err != nil {
+				return false, err
 			}
 		}
 	}
-	err := RunWorkflowWithCancelSignal(context.Background(), w.wf)
-	if w.wf.Logger != nil {
-		for _, trace := range w.wf.Logger.ReadSerialPortLogs() {
+	err = RunWorkflowWithCancelSignal(wf, w.cancel)
+	if wf.Logger != nil {
+		for _, trace := range wf.Logger.ReadSerialPortLogs() {
 			w.logger.Trace(trace)
 		}
 	}
 	if err != nil {
-		PostProcessDErrorForNetworkFlag(w.env.Tool.HumanReadableName, err, w.env.Network, w.wf)
+		PostProcessDErrorForNetworkFlag(w.env.Tool.HumanReadableName, err, w.env.Network, wf)
+		for _, hook := range w.hooks {
+			preHook, isPreHook := hook.(WorkflowPostHook)
+			if isPreHook {
+				wantRetry := false
+				wantRetry, err = preHook.PostRunHook(err)
+				retry = retry || wantRetry
+			}
```

Comment:
This is hard to understand: preHook is run as postHook? Maybe add some comments to clarify the logic here.

Reply:
Shoot, the variable names are incorrect for the type. Updated.